authenticator/authenticator.go

```diff
@@ -0,0 +1,178 @@
+		return nil
+	}
+}
+
+// WithApiAudience sets the API audience for access token verification.
+func WithApiAudience(aud string) Option {
+	return func(a *Authenticator) error {
+		a.ApiAudience = aud
+		return nil
+	}
+}
+
+// WithLogger sets the logger for the authenticator.
+func WithLogger(log *logger.Logger) Option {
+	return func(a *Authenticator) error {
+		a.log = log.With(lctx.Str("component", "authenticator"))
+		return nil
+	}
+}
+
+func New(ctx context.Context, domain string, opts ...Option) (*Authenticator, error) {
+	// Create OIDC provider
+	provider, err := oidc.NewProvider(ctx, "https://"+domain+"/")
+	if err != nil {
+		return nil, fmt.Errorf("failed to create OIDC provider: %w", err)
+	}
+
+	// Configure OAuth2
+	auth := &Authenticator{
+		Provider: provider,
+		Config: oauth2.Config{
+			Endpoint: provider.Endpoint(),
+			Scopes:   []string{oidc.ScopeOpenID, "profile", "email"},
+		},
+	}
+
+	for _, opt := range opts {
+		if err := opt(auth); err != nil {
+			return nil, fmt.Errorf("failed to apply options: %w", err)
+		}
+	}
+
+	return auth, nil
+}
+
+// verifyToken verifies the ID token and returns the parsed token.
+func (a *Authenticator) verifyToken(ctx context.Context, token *oauth2.Token) (*oidc.IDToken, error) {
+	rawIDToken, ok := token.Extra("id_token").(string)
+	if !ok {
+		return nil, ErrNoIDToken
+	}
+
+	oidcConfig := &oidc.Config{
+		ClientID: a.Config.ClientID,
+	}
+	verifier := a.Provider.Verifier(oidcConfig)
+	return verifier.Verify(ctx, rawIDToken)
+}
+
+// verifyAccessToken verifies an access token.
+func (a *Authenticator) verifyAccessToken(ctx context.Context, token *oauth2.Token) (*oidc.IDToken, error) {
+	oidcConfig := &oidc.Config{
+		ClientID: a.ApiAudience,
+	}
+	verifier := a.Provider.Verifier(oidcConfig)
+
+	return verifier.Verify(ctx, token.AccessToken)
+}
+
+// revokeToken sends a request to Auth0 to revoke the token.
+func (a *Authenticator) revokeToken(ctx context.Context, token string) error {
+	domain := a.getDomain()
+	revokeURL := fmt.Sprintf("https://%s/oauth/revoke", domain)
+
+	// Prepare the request body
+	form := url.Values{}
+	form.Add("client_id", a.Config.ClientID)
+	form.Add("client_secret", a.Config.ClientSecret)
+	form.Add("token", token)
+
+	// Create and send the request
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, revokeURL, strings.NewReader(form.Encode()))
+	if err != nil {
+		return fmt.Errorf("failed to create revocation request: %w", err)
+	}
+
+	req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return fmt.Errorf("failed to send revocation request: %w", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		body, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return fmt.Errorf("failed to read revocation response: %w", err)
+		}
+		return fmt.Errorf("failed to revoke token with status %d: %s", resp.StatusCode, string(body))
```

Comment:
Error strings should be shorter since they will be chained/wrapped

```suggestion
		return fmt.Errorf("sending revocation request: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, err := io.ReadAll(resp.Body)
		if err != nil {
			return fmt.Errorf("reading revocation response: %w", err)
		}
		return fmt.Errorf("revoking token with status %d: %s", resp.StatusCode, string(body))
```